File: management-ui/backend/app/schemas/metrics.py
from pydantic import BaseModel, Field
from typing import List, Dict, Optional
from datetime import datetime
from enum import Enum


class AlertSeverity(str, Enum):
    INFO = "info"
    WARNING = "warning"
    CRITICAL = "critical"


class AlertType(str, Enum):
    HIGH_CPU = "high_cpu"
    HIGH_MEMORY = "high_memory"
    CONTAINER_RESTART = "container_restart"
    HEALTH_CHECK_FAILED = "health_check_failed"
    OOM_KILLED = "oom_killed"
    NETWORK_ERROR = "network_error"


class ContainerMetrics(BaseModel):
    """Real-time metrics for a single container."""
    service_name: str
    container_id: str
    timestamp: datetime
    cpu_percent: float = Field(ge=0, description="CPU usage percentage")
    memory_usage: int = Field(ge=0, description="Memory usage in bytes")
    memory_limit: int = Field(ge=0, description="Memory limit in bytes")
    memory_percent: float = Field(ge=0, le=100, description="Memory usage percentage")
    network_rx_bytes: int = Field(ge=0, description="Network bytes received")
    network_tx_bytes: int = Field(ge=0, description="Network bytes transmitted")
    network_rx_rate: float = Field(ge=0, description="Network receive rate bytes/sec")
    network_tx_rate: float = Field(ge=0, description="Network transmit rate bytes/sec")
    block_read_bytes: int = Field(ge=0, default=0)
    block_write_bytes: int = Field(ge=0, default=0)


class MetricsSnapshot(BaseModel):
    """Aggregated metrics snapshot for storage."""
    service_name: str
    timestamp: datetime
    cpu_avg: float
    cpu_max: float
    memory_avg: float
    memory_max: float
    network_rx_rate_avg: float
    network_tx_rate_avg: float


class MetricsHistoryResponse(BaseModel):
    """Response for metrics history query."""
    service_name: str
    start_time: datetime
    end_time: datetime
    data_points: List[MetricsSnapshot]
    granularity_seconds: int = 60


class HealthAlert(BaseModel):
    """Health alert notification."""
    id: str
    service_name: str
    alert_type: AlertType
    severity: AlertSeverity
    message: str
    timestamp: datetime
    acknowledged: bool = False
    details: Optional[Dict] = None


class DiagnosticResult(BaseModel):
    """Result of diagnostic check."""
    service_name: str
    issue: str
    description: str
    severity: AlertSeverity
    recommendation: str
    auto_fixable: bool = False


class DiagnosticsResponse(BaseModel):
    """Response for diagnostics endpoint."""
    service_name: str
    issues: List[DiagnosticResult]
    restart_recommended: bool
    last_restart: Optional[datetime] = None
    uptime_seconds: Optional[int] = None


class AllMetricsResponse(BaseModel):
    """Response for all services metrics."""
    metrics: List[ContainerMetrics]
    timestamp: datetime

File: management-ui/frontend/src/types/metrics.ts
export type AlertSeverity = 'info' | 'warning' | 'critical';
export type AlertType = 'high_cpu' | 'high_memory' | 'container_restart' | 'health_check_failed' | 'oom_killed' | 'network_error';

export interface ContainerMetrics {
  service_name: string;
  container_id: string;
  timestamp: string;
  cpu_percent: number;
  memory_usage: number;
  memory_limit: number;
  memory_percent: number;
  network_rx_bytes: number;
  network_tx_bytes: number;
  network_rx_rate: number;
  network_tx_rate: number;
  block_read_bytes?: number;
  block_write_bytes?: number;
}

export interface MetricsSnapshot {
  service_name: string;
  timestamp: string;
  cpu_avg: number;
  cpu_max: number;
  memory_avg: number;
  memory_max: number;
  network_rx_rate_avg: number;
  network_tx_rate_avg: number;
}

export interface MetricsHistoryResponse {
  service_name: string;
  start_time: string;
  end_time: string;
  data_points: MetricsSnapshot[];
  granularity_seconds: number;
}

export interface HealthAlert {
  id: string;
  service_name: string;
  alert_type: AlertType;
  severity: AlertSeverity;
  message: string;
  timestamp: string;
  acknowledged: boolean;
  details?: Record<string, unknown>;
}

export interface DiagnosticResult {
  service_name: string;
  issue: string;
  description: string;
  severity: AlertSeverity;
  recommendation: string;
  auto_fixable: boolean;
}

export interface DiagnosticsResponse {
  service_name: string;
  issues: DiagnosticResult[];
  restart_recommended: boolean;
  last_restart: string | null;
  uptime_seconds: number | null;
}

export interface AllMetricsResponse {
  metrics: ContainerMetrics[];
  timestamp: string;
}

export interface MetricsWebSocketMessage {
  type: 'metrics' | 'error';
  timestamp?: string;
  data?: ContainerMetrics | Record<string, ContainerMetrics>;
  message?: string;
}
